### ISSUE TITLE
錯誤提示顯示方式制定

### ISSUE THREAD

**Issue by diabloevagto** [TaiBIF/camera-trap-vueapp#26]
@cjk-halodule 
有關之前討論到 API error 的提示,
是前端依照 status code 顯示對應的文字資訊給使用者對嗎？
所以應該可以統一都用跳出視窗顯示即可？

_Originally posted by @oahehc in https://github.com/TaiBIF/camera-trap-vueapp/pull/25#discussion_r277133795

**Comment by cjk-halodule** [comment]
是的。有和 @kelp404 討論到要列出後端會回傳的錯誤訊息及狀態的列表，然後再對應給前端用來和使用者說明的文字。我們先在[這裡](https://github.com/TaiBIF/camera-trap-api/wiki/Status-Code-and-Error-Code)共筆好了...

**Comment by cjk-halodule** [comment]
@diabloevagto @oahehc @f2ejack 

我把目前幾個前端反映狀態給使用者的類型另外整理在一頁：
[TaiBIF Camera Trap wireframe v20190513 (dragged).pdf](https://github.com/TaiBIF/camera-trap-vueapp/files/3172328/TaiBIF.Camera.Trap.wireframe.v20190513.dragged.pdf)

目前在 camera-trap-api 中掃到 400 的錯誤先列在**[這裡](https://github.com/TaiBIF/camera-trap-api/wiki/Status-Code-and-Error-Code)**，看看這對於實作有否幫助？

@kelp404 其中的文字可能要請您幫忙看看是不是 makes sense？

然後，有些錯誤可能 logging 起來比和使用者詳細說明重要，再請大家處理到時提議了。

**Comment by kelp404** [comment]
我覺得錯誤提示由前端的表單驗證處理，如果需要後端邏輯驗證的話加開 API，例如新增計畫成員前檢查帳號是否存在。
目前 API 的錯誤提示都是給開發者用的，而且有些沒有說明原因，所以表格中沒有 message。
但這些錯誤都是不該發生的，例如使用者在一個不存在的計畫新增相機位置、對不存在的相機位置上傳 csv...。

等下一階段看要不要制定錯誤代碼，因為一個錯誤原因就要一個錯誤代碼，這會非常花時間。

錯誤都有紀錄，有 error stack，只要版號一致就能對得起來是哪一行報錯。
<img width="938" alt="螢幕快照 2019-05-13 下午6 29 25" src="https://user-images.githubusercontent.com/1680868/57615013-3d50ac80-75ad-11e9-9a61-17560755e334.png">

**Comment by cjk-halodule** [comment]
剛在表格中將前端明顯是應由表單驗證的錯誤加上 FV。
表格中沒有資訊的項目如果是前端必須回應使用者的情況，目前就提示使用者可以回報給系統管理員。

也針對 @kelp404 的建議開了一個下階段的待辦事項。

**Comment by f2ejack** [comment]
> 我覺得錯誤提示由前端的表單驗證處理，如果需要後端邏輯驗證的話加開 API，例如新增計畫成員前檢查帳號是否存在。
> 目前 API 的錯誤提示都是給開發者用的，而且有些沒有說明原因，所以表格中沒有 message。
> 但這些錯誤都是不該發生的，例如使用者在一個不存在的計畫新增相機位置、對不存在的相機位置上傳 csv...。
> 
> 等下一階段看要不要制定錯誤代碼，因為一個錯誤原因就要一個錯誤代碼，這會非常花時間。
> 
> 錯誤都有紀錄，有 error stack，只要版號一致就能對得起來是哪一行報錯。
> <img alt="螢幕快照 2019-05-13 下午6 29 25" width="938" src="https://user-images.githubusercontent.com/1680868/57615013-3d50ac80-75ad-11e9-9a61-17560755e334.png">
的確, 至少在表單的欄位部分必須由前端來做第一步的驗證, 例如: 必填欄位未填寫, 資料格式錯誤, ... 等.

**Comment by f2ejack** [comment]
> 剛在表格中將前端明顯是應由表單驗證的錯誤加上 FV。
> 表格中沒有資訊的項目如果是前端必須回應使用者的情況，目前就提示使用者可以回報給系統管理員。
> 
> 也針對 @kelp404 的建議開了一個下階段的待辦事項。

"回報給系統管理員"是指連結到"聯絡我們"的那個頁面嗎?

**Comment by cjk-halodule** [comment]
> "回報給系統管理員"是指連結到"聯絡我們"的那個頁面嗎?

是的。

**Comment by oahehc** [comment]
前端做表單驗證, 也還是需要定義 API error 時要顯示給 user 的訊息, 最少要 by status 提供對應資訊, 類似以下：
```
  switch (status) {
    case '400':
      return '操作失敗，資料內容有誤。';
    case '401':
      return '操作失敗，請先登入。';
    case '403':
      return '操作失敗，您的帳號無權限操作此步驟';
    default:
      return '操作失敗，請再試一次';
  }
```

**Comment by f2ejack** [comment]
通常都會有一個頁面是針對處理這類訊息的, 例如很多公司花巧思做的 404 - Page Not Found 頁面.

**Comment by oahehc** [comment]
no, 404 - Page Not Found 跟 API error handle 不一樣,
404 是 user get 一個不存在的網址。
而這邊要處理的 API error handle 是前端 call API 但 BE server 執行失敗,
此時需要顯示對應資訊讓 user 資料操作失敗, 以及應該如何處理。

**Comment by cjk-halodule** [comment]
﻿﻿﻿﻿https://github.com/TaiBIF/camera-trap-api/issues/168#issuecomment-495645850

**Comment by f2ejack** [comment]
持續整理後端所回傳的錯誤訊息, 已針對 handler/file-handler.js 檔案進行了更新及補充, 其中 Frontend 一欄標註 WR (wait for reviewing) 的訊息內容, 需要請 @cjk-halodule 協助 review.
https://github.com/TaiBIF/camera-trap-api/wiki/Status-Code-and-Error-Code

**Comment by moogoo78** [comment]
要處理，如果同一個 CSV 檔案，遇到資料中有不同相機位置時，需要回傳讓使用者能了解的錯誤訊息

**Comment by f2ejack** [comment]
**交接事項:**

- 目前已經整理完有關前端呼叫 API 後, 回應狀態為 400 的 Error Messages 都整理在這裡 (以下)：
[Status-Code-and-Error-Code](https://github.com/TaiBIF/camera-trap-api/wiki/Status-Code-and-Error-Code)
- @cjk-halodule 之前已經把目前幾個前端反映狀態給使用者的類型另外整理在一頁 (如下)：
[TaiBIF Camera Trap wireframe v20190513 (dragged).pdf](https://github.com/TaiBIF/camera-trap-vueapp/files/3172328/TaiBIF.Camera.Trap.wireframe.v20190513.dragged.pdf)
- 除了 Status 為 400 的 Messages 需要以代碼方式 (這是我的想法, 僅供參考) 回傳給前端處理之外, 其它的狀態請與 @moogoo78 討論
- 前端開發者必須透過這份整理出來的 Error Messages 來顯示相關的訊息 (請掃過所有的頁面, 看那裡需要加上這些互動的訊息), 以及其呈現的形式為何 (e.g., Modal, Inline Message, ...)？
- 需考慮到未來 i18n 的需求
